topology/init.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topology
+
+import "errors"
+
+var errNilMapOptions = errors.New("nil map options")
+
+type topologyInitializer struct {
+	opts Options
+}
+
+// NewInitializer returns a topology Initializer
+func NewInitializer(opts Options) Initializer {
+	return topologyInitializer{opts}
+}
+
+func (c topologyInitializer) Init() (Topology, error) {
+	t, err := newDynamicTopology(c.opts)
+	if err == nil {
+		return t, nil
+	}
+	// Failed to create dynamic topology, try to fail back to static topology
```

Comment:
Hm doesn't seem like a very explicit strategy.  I'd be in prefer of creating both a static initializer and a dynamic initializer and each resides in its own package.